### ISSUE TITLE
Add html content check condition

### ISSUE THREAD

**Issue by whysthatso** [TwinProduction/gatus#68]
As mentioned in my previous issue #61 i'm currently using healthchecks, and the last feature i would need is some generic content regex / string presence check. now, i might be missing something everybody already knows, since you have this BODY placeholder, and maybe that's just it.

anyway, it seems that the process basically would mean gatus curls the site content and runs a regex or literal string against it, and when it finds it, test passes.

what do you think?

in terms of performance, i think that this is actually the most heavy that healthchecks is doing, but already now i'm running all the tests in parallel, and gatus outperforms healthchecks by magnitudes. so nice.

**Comment by TwinProduction** [comment]
This is technically already possible by leveraging the placeholder `[BODY]` with the function `pat` (short for pattern). 

I did just find a small issue with `pat` due to backslashes, which I'll fix in a bit, but let me know if this is sufficient.

Here's an example:

```yaml
services:
  - name: example
    url: "https://twinnation.org/"
    interval: 5m
    conditions:
      - "[BODY] == pat(*DOCTYPE*)"
```

**Comment by whysthatso** [comment]
ok, interesting. i've put your example in with one of my sites, literally checking for that string DOCTYPE, but for some reason it returns false. weirdly enough on the status page i get a green checkmark, but the check details look like this:

```
Timestamp:
2021-01-06 16:09:59
Response time:
42ms
Conditions:
X ~ [BODY] == pat(*DOCTYPE*) ( == pat(*DOCTYPE*))
```
same with the literal string:

```
Timestamp:
2021-01-06 16:14:02
Response time:
43ms
Conditions:
X ~ [BODY] == pat(DOCTYPE) ( == pat(DOCTYPE))
```

could i run this test somehow from the command line, as a go oneliner, just to see some more details, how the BODY content looks like etc?

**Comment by whysthatso** [comment]
sidenote: i meant to say i use cabot, not healthchecks for this kind of work previously.

**Comment by TwinProduction** [comment]
@whysthatso Have you run this with the latest version (`v1.11.0`)? 

I fixed an issue caused with pattern caused by backslashes.
Originally, the `pat` function wasn't intended for HTML, so I hadn't tested with HTML content, but after doing so, I found out that issue and resolved it.

**Comment by whysthatso** [comment]
i'm running this now on the latest 1.11.2, situation unchanged.

**Comment by whysthatso** [comment]
again, is there any way to run gatus in a debug mode to see the content of BODY?

**Comment by TwinProduction** [comment]
@whysthatso Technically, if a condition does not succeed, it should automatically print the content of the body.

Here's an example:
```yaml
services:
  - name: example
    url: "http://example.org/"
    interval: 5m
    conditions:
      - "[BODY] == pat(wow*)"
```

Logs: 
```
2021/01/15 09:27:32 [watchdog][monitor] Monitored serviceName=example; success=false; errors=0; requestDuration=53ms; responseBody=<!doctype html>
<html>
<head>
    <title>Example Domain</title>

    <meta charset="utf-8" />
    <meta http-equiv="Content-type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type="text/css">
    body {
        background-color: #f0f0f2;
        margin: 0;
        padding: 0;
        font-family: -apple-system, system-ui, BlinkMacSystemFont, "Segoe UI", "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
        
    }
    div {
        width: 600px;
        margin: 5em auto;
        padding: 2em;
        background-color: #fdfdff;
        border-radius: 0.5em;
        box-shadow: 2px 3px 7px 2px rgba(0,0,0,0.02);
    }
    a:link, a:visited {
        color: #38488f;
        text-decoration: none;
    }
    @media (max-width: 700px) {
        div {
            margin: 0 auto;
            width: auto;
        }
    }
    </style>    
</head>

<body>
<div>
    <h1>Example Domain</h1>
    <p>This domain is for use in illustrative examples in documents. You may use this
    domain in literature without prior coordination or asking for permission.</p>
    <p><a href="https://www.iana.org/domains/example">More information...</a></p>
</div>
</body>
</html>
```

**Comment by whysthatso** [comment]
yes, i've seen something like that briefly before, however at the moment i get only this:

`[Condition][evaluate] Condition '[BODY] == pat(*DOCTYPE*)' did not succeed because ' == pat(*DOCTYPE*)' is false`

**Comment by whysthatso** [comment]
okay, my bad. i've pointed the test to a wrong domain... nevermind, i see the fail output now.

however, 

 X ~ [BODY] == pat(<!DOCTYPE html>) (<!DOCTYPE html> <html lang="et"> <head> <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />.... == pat(<!DOCTYPE html>))

**Comment by whysthatso** [comment]
i've tried a really universal term now

```
"[BODY] == pat(*div*)"
```

still all fails...

**Comment by TwinProduction** [comment]
@whysthatso Is there any chance you could provide me with the body of the page? If there's any confidential information, feel free to exclude it.

If I can reproduce it, I'll fix it, but with the websites I've tried, I haven't run into this issue.

Once I have the body, I'll create a test for it and trim down the body to as small as possible until I stop getting the issue; that'll allow me to pinpoint what exactly the issue is.

i.e.
```go
func TestCondition_evaluateWithBodyHTMLPattern(t *testing.T) {
	var html = `<!DOCTYPE html> <html lang="et"> <head> <meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />....`
	condition := Condition("[BODY] == pat(*head*)")
	result := &Result{Body: []byte(html)}
	condition.evaluate(result)
	if !result.ConditionResults[0].Success {
		t.Errorf("Condition '%s' should have been a success", condition)
	}
	expectedConditionDisplayed := "[BODY] == pat(*head*)"
	if result.ConditionResults[0].Condition != expectedConditionDisplayed {
		t.Errorf("Condition '%s' should have resolved to '%s', got '%s'", condition, expectedConditionDisplayed, result.ConditionResults[0].Condition)
	}
}
```

**Comment by TwinProduction** [comment]
@whysthatso Actually, I think I may have found the issue.

Could re-pull the `latest` docker tag and try again?

**Comment by whysthatso** [comment]
did that, but to no avail. i've actually pruned everything, set up from scratch and only ran your test example, with debug on, this is the config:

```

debug: true
metrics: true
alerting:
  mattermost:
    webhook-url: "https://url.to.mattermost/hooks/<secret>"
    insecure: false

services:
  - name: example
    url: "https://twinnation.org/"
    interval: 30s
    alerts:
      - type: mattermost
        enabled: true
        description: "healthcheck failed 3 times in a row"
        send-on-resolved: true
    conditions:
      - "[BODY] == pat(*DOCTYPE*)"
```

this is my compose file:

```
version: "3"
services:
  gatus:
    image: twinproduction/gatus:latest
    container_name: gatus
    ports:
      - "39425:8080"
    volumes:
      - ./config.yaml:/config/config.yaml
```

this is the output of `docker-compose up`

```
root@bbox004:/opt/gatus# docker-compose up 
Pulling gatus (twinproduction/gatus:latest)...
latest: Pulling from twinproduction/gatus
bad8fffb4eda: Pull complete
555aa233bb34: Pull complete
a70b215ed0fb: Pull complete
d2fffcad6470: Pull complete
Digest: sha256:c7ebb4332495c72ad3b7fbe29708203072f1e32d939a6745ace7231f53c48381
Status: Downloaded newer image for twinproduction/gatus:latest
Creating gatus ... 
Creating gatus ... done
Attaching to gatus
gatus    | 2021/01/16 15:56:03 [config][Load] Reading configuration from configFile=config/config.yaml
gatus    | 2021/01/16 15:56:03 [config][validateAlertingConfig] configuredProviders=[mattermost]; ignoredProviders=[slack messagebird twilio pagerduty custom]
gatus    | 2021/01/16 15:56:03 [config][validateServicesConfig] Validating service 'example'
gatus    | 2021/01/16 15:56:03 [config][validateServicesConfig] Validated 1 services
gatus    | 2021/01/16 15:56:03 [watchdog][monitor] Monitoring serviceName=example
gatus    | 2021/01/16 15:56:03 [controller][Handle] Listening on 0.0.0.0:8080/
gatus    | 2021/01/16 15:56:04 [watchdog][monitor] Monitored serviceName=example; success=false; errors=0; requestDuration=527ms; responseBody=<!DOCTYPE html>
gatus    | <html lang="en">
gatus    | <head>
gatus    |     <meta charset="UTF-8" />
gatus    |     <title>TwiNNatioN</title>
gatus    |     
gatus    |     <meta property="og:type" content="website" />
gatus    |     <meta property="og:site_name" content="TwiNNatioN" />
gatus    |     <meta name="twitter:card" content="summary" />
gatus    |     <meta name="twitter:title" property="og:title" content="TwiNNatioN" />
gatus    |     
gatus    |     
gatus    |     <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
gatus    |     <meta name="viewport" content="width=device-width, initial-scale=1" />
gatus    |     <meta name="theme-color" content="#DD0000" />
gatus    |     <link rel="manifest" href="/manifest.json" />
gatus    |     <link rel="shortcut icon" href="/favicon.ico" />
gatus    |     <link rel="stylesheet" href="/css/bootstrap.min.css" />
gatus    |     <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.0.9/css/all.css" />
gatus    |     <link href="https://fonts.googleapis.com/css?family=Maven+Pro" rel="stylesheet" />
gatus    |     <script data-ad-client="ca-pub-1050220987226942" async src="https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js"></script>
gatus    | </head>
gatus    | <body>
gatus    |     <script>(function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){(i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)})(window,document,'script','https://www.google-analytics.com/analytics.js','ga');ga('create', 'UA-76489219-1', 'auto');ga('send', 'pageview');</script>
gatus    |     <noscript><div class="text-center"><i>You need to enable JavaScript to run this app.</i></div></noscript>
gatus    | 
gatus    |     <div id="root">
gatus    |         
gatus    |         <div class="container bg-light rounded p-2 my-3">
gatus    |             <h1>TwiNNatioN</h1>
gatus    |             
gatus    |         </div>
gatus    |         
gatus    |     </div>
gatus    | 
gatus    |     <script src="/vendor.bundle.js"></script>
gatus    |     <script src="/main.bundle.js"></script>
gatus    | </body>
gatus    | </html>
gatus    | 2021/01/16 15:56:04 [watchdog][monitor] Waiting for interval=1m0s before monitoring serviceName=example again
```

there's also an issue with notifying mattermost, but i will open another issue for that if need be:

```
gatus    | 2021/01/16 16:02:23 [watchdog][handleAlertsToTrigger] Sending mattermost alert because alert with description='healthcheck failed 3 times in a row' has been TRIGGERED
gatus    | 2021/01/16 16:02:23 [watchdog][handleAlertsToTrigger] Ran into error sending an alert: call to provider alert returned status code 400: {"id":"model.incoming_hook.parse_data.app_error","message":"Unable to parse incoming data.","detailed_error":"","request_id":"znoajjaje7gsuphp1yry9tkbcw","status_code":400}
```

**Comment by TwinProduction** [comment]
@whysthatso I was able to reproduce the issue by pulling `twinproduction/gatus:latest`, and it turns out that yesterday's Docker build job got stuck, which is why the issue was still present.

I've retried the Docker build task, and the `latest` tag has now been replaced by the one with fix I mentioned yesterday.

I apologize for the inconvenience, it should work this time (excluding the mattermost issue, which I'll look at separately).

**Comment by whysthatso** [comment]
yey, you did it. it works now nice.

**Comment by TwinProduction** [comment]
Glad to hear that.
Since `pat` fulfills your use case, would you be fine with me closing this issue?

**Comment by TwinProduction** [comment]
The fix for the pattern function has been released in [v1.12.0](https://github.com/TwinProduction/gatus/releases/tag/v1.12.0)
